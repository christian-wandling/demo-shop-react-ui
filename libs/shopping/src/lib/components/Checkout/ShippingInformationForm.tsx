import { Form, FormInput, FormSelect } from '@demo-shop-react-ui/shared';
import { ShippingInformationProps } from '../../models/shippingInformationProps';
import { useEffect, useState } from 'react';
import { countrySelectOptions } from '../../constants/countrySelectOptions';
import { ShippingInformationFormData } from '../../models/shippingInformationFormSchema';

export const ShippingInformationForm = ({ methods, onUpdateUser }: ShippingInformationProps) => {
  const BREAKPOINT_MD = 768;
  const [shippingExtended, setShippingExtended] = useState(false);

  useEffect(() => {
    setShippingExtended(window.innerWidth >= BREAKPOINT_MD);

    const handleResize = () => {
      setShippingExtended(window.innerWidth >= BREAKPOINT_MD);
    };

    window.addEventListener('resize', handleResize);
    return () => window.removeEventListener('resize', handleResize);
  }, []);

  const toggleShippingExtended = () => {
    setShippingExtended(!shippingExtended);
  };

  const updateButtonEnabled = methods.formState.isValid && methods.formState.isDirty;

  return (
    <Form<ShippingInformationFormData> methods={methods} onSubmit={onUpdateUser} className="user-form">
      <div className="space-y-8">
        <div
          className="flex justify-between items-center p-4 bg-gray-100 cursor-pointer lg:pointer-events-none"
          onClick={toggleShippingExtended}>
          <h2 className="text-base font-semibold leading-7 text-gray-900">Shipping information</h2>
          <svg
            xmlns="http://www.w3.org/2000/svg"
            className={`lg:hidden w-5 h-5 transition-transform duration-300 ${shippingExtended ? 'rotate-180' : ''}`}
            fill="none"
            viewBox="0 0 24 24"
            stroke="currentColor">
            <path strokeLinecap="round" strokeLinejoin="round" strokeWidth="2" d="M19 9l-7 7-7-7" />
          </svg>
        </div>
        <div
          className={`mt-4 lg:max-h-[1000px] grid md:grid-cols-8 gap-x-6 gap-y-4 transition-all duration-500 ease-in-out overflow-hidden ${
            shippingExtended ? 'max-h-[1000px]' : 'max-h-0'
          }`}>
          <FormInput
            name="firstname"
            label="Firstname"
            disabled
            required
            className="md:col-span-4"
            data-testid="checkout-form-element-firstname"
          />
          <FormInput
            name="lastname"
            label="Firstname"
            disabled
            required
            className="md:col-span-4"
            data-testid="checkout-form-element-lastname"
          />
          <FormInput
            name="email"
            label="Email"
            disabled
            required
            className="md:col-span-4"
            data-testid="checkout-form-element-email"
          />
          <FormInput name="phone" label="Phone" className="md:col-span-4" data-testid="checkout-form-element-phone" />
          <FormSelect
            name="address.country"
            label="Country"
            options={countrySelectOptions}
            required
            className="md:col-span-4"
            data-testid="checkout-form-element-country"
          />
          <FormInput
            name="address.street"
            label="Street"
            required
            className="md:col-span-5"
            data-testid="checkout-form-element-street"
          />
          <FormInput
            name="address.apartment"
            label="Apartment"
            required
            className="md:col-span-3"
            data-testid="checkout-form-element-apartment"
          />
          <FormInput
            name="address.city"
            label="City"
            required
            className="md:col-span-3"
            data-testid="checkout-form-element-city"
          />
          <FormInput
            name="address.region"
            label="Region"
            className="md:col-span-3"
            data-testid="checkout-form-element-region"
          />
          <FormInput
            name="address.zip"
            label="Zip"
            required
            className="md:col-span-2"
            data-testid="checkout-form-element-zip"
          />

          <button
            disabled={!updateButtonEnabled}
            className={`btn-update md:col-span-full float-right m-2 mt-6 lg:max-w-60 flex flex-1 items-center justify-center rounded-md border border-transparent px-8 py-3 text-base font-medium ${
              !updateButtonEnabled
                ? 'bg-gray-200 text-gray-800 pointer-events-none'
                : 'bg-indigo-600 text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2'
            }`}
            data-testid="checkout-update-shipping-information-button">
            Update
          </button>
        </div>
      </div>
    </Form>
  );
};
