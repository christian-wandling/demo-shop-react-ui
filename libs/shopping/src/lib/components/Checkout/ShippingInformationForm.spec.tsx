import { beforeEach, describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { ShippingInformationForm } from './ShippingInformationForm';

Object.defineProperty(window, 'innerWidth', {
  writable: true,
  configurable: true,
  value: 1024,
});

vi.mock('@demo-shop-react-ui/shared', () => ({
  Form: ({ children, onSubmit }: any) => (
    <form
      data-testid="mock-form"
      onSubmit={e => {
        e.preventDefault();
        onSubmit();
      }}>
      {children}
    </form>
  ),
  FormInput: ({ name, label, disabled, required, className }: any) => (
    <div data-testid={`form-input-${name}`} className={className}>
      <label>
        {label}
        {required && '*'}
      </label>
      <input name={name} disabled={disabled} required={required} data-testid={`input-${name}`} />
    </div>
  ),
  FormSelect: ({ name, label, options, required, className }: any) => (
    <div data-testid={`form-select-${name}`} className={className}>
      <label>
        {label}
        {required && '*'}
      </label>
      <select name={name} required={required} data-testid={`select-${name}`}>
        {options.map((option: any) => (
          <option key={option.value} value={option.value}>
            {option.label}
          </option>
        ))}
      </select>
    </div>
  ),
}));

vi.mock('../../constants/countrySelectOptions', () => ({
  countrySelectOptions: [
    { value: 'US', label: 'United States' },
    { value: 'CA', label: 'Canada' },
  ],
}));

describe('ShippingInformationForm', () => {
  const mockMethods: any = {
    formState: {
      isValid: false,
      isDirty: false,
    },
    control: {},
    handleSubmit: vi.fn(callback => (data: any) => callback(data)),
  };

  const mockOnUpdateUser = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('matches the snapshot', () => {
    const { baseElement } = render(<ShippingInformationForm methods={mockMethods} onUpdateUser={mockOnUpdateUser} />);

    expect(baseElement).toMatchSnapshot();
  });

  it('renders the form with correct title', () => {
    render(<ShippingInformationForm methods={mockMethods} onUpdateUser={mockOnUpdateUser} />);

    expect(screen.getByText('Shipping information')).toBeTruthy();
  });

  it('expands on mobile view when header is clicked', () => {
    (window as any).innerWidth = 500;

    render(<ShippingInformationForm methods={mockMethods} onUpdateUser={mockOnUpdateUser} />);

    const formContent = screen.getByTestId('mock-form').parentElement;
    expect(formContent?.querySelector('div[class*="max-h-0"]')).toBeTruthy();

    fireEvent.click(screen.getByText('Shipping information'));

    expect(formContent?.querySelector('div[class*="max-h-[1000px]"]')).toBeTruthy();
  });

  it('renders all form fields with correct attributes', () => {
    render(<ShippingInformationForm methods={mockMethods} onUpdateUser={mockOnUpdateUser} />);

    expect(screen.getByTestId('form-input-firstname')).toBeTruthy();
    expect(screen.getByTestId('form-input-lastname')).toBeTruthy();
    expect(screen.getByTestId('form-input-email')).toBeTruthy();
    expect(screen.getByTestId('form-input-phone')).toBeTruthy();
    expect(screen.getByTestId('form-select-address.country')).toBeTruthy();
    expect(screen.getByTestId('form-input-address.street')).toBeTruthy();
    expect(screen.getByTestId('form-input-address.apartment')).toBeTruthy();
    expect(screen.getByTestId('form-input-address.city')).toBeTruthy();
    expect(screen.getByTestId('form-input-address.region')).toBeTruthy();
    expect(screen.getByTestId('form-input-address.zip')).toBeTruthy();

    expect((screen.getByTestId('input-firstname') as HTMLInputElement).disabled).toBeTruthy();
    expect((screen.getByTestId('input-lastname') as HTMLInputElement).disabled).toBeTruthy();
    expect((screen.getByTestId('input-email') as HTMLInputElement).disabled).toBeTruthy();
  });

  it('disables update button when form is not valid or dirty', () => {
    render(<ShippingInformationForm methods={mockMethods} onUpdateUser={mockOnUpdateUser} />);

    const updateButton: HTMLButtonElement = screen.getByText('Update');
    expect(updateButton.disabled).toBe(true);
    expect(updateButton.classList).toContain('bg-gray-200');
  });

  it('enables update button when form is valid and dirty', () => {
    const validMethods = {
      ...mockMethods,
      formState: {
        isValid: true,
        isDirty: true,
      },
    };

    render(<ShippingInformationForm methods={validMethods} onUpdateUser={mockOnUpdateUser} />);

    const updateButton: HTMLButtonElement = screen.getByText('Update');
    expect(updateButton.disabled).toBe(false);
    expect(updateButton.classList).toContain('bg-indigo-600');
  });

  it('calls onUpdateUser when form is submitted', () => {
    const validMethods = {
      ...mockMethods,
      formState: {
        isValid: true,
        isDirty: true,
      },
    };

    render(<ShippingInformationForm methods={validMethods} onUpdateUser={mockOnUpdateUser} />);

    const form = screen.getByTestId('mock-form');
    fireEvent.submit(form);

    expect(mockOnUpdateUser).toHaveBeenCalled();
  });

  it('is expanded by default on desktop', () => {
    (window as any).innerWidth = 1024;

    render(<ShippingInformationForm methods={mockMethods} onUpdateUser={mockOnUpdateUser} />);

    const formContent = screen.getByTestId('mock-form').parentElement;
    expect(formContent?.querySelector('div[class*="max-h-[1000px]"]')).toBeTruthy();
  });

  it('has correct responsive classes', () => {
    render(<ShippingInformationForm methods={mockMethods} onUpdateUser={mockOnUpdateUser} />);

    expect(screen.getByTestId('form-input-firstname').classList).toContain('md:col-span-4');
    expect(screen.getByTestId('form-input-address.street').classList).toContain('md:col-span-5');
    expect(screen.getByTestId('form-input-address.zip').classList).toContain('md:col-span-2');
  });
});
