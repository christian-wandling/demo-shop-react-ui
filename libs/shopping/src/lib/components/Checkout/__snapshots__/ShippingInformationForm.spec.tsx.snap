// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShippingInformationForm > matches the snapshot 1`] = `
<body>
  <div>
    <form
      data-testid="mock-form"
    >
      <div
        class="space-y-8"
      >
        <div
          class="flex justify-between items-center p-4 bg-gray-100 cursor-pointer lg:pointer-events-none"
        >
          <h2
            class="text-base font-semibold leading-7 text-gray-900"
          >
            Shipping information
          </h2>
          <svg
            class="lg:hidden w-5 h-5 transition-transform duration-300 rotate-180"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 9l-7 7-7-7"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <div
          class="mt-4 lg:max-h-[1000px] grid md:grid-cols-8 gap-x-6 gap-y-4 transition-all duration-500 ease-in-out overflow-hidden max-h-[1000px]"
        >
          <div
            class="md:col-span-4"
            data-testid="form-input-firstname"
          >
            <label>
              Firstname
              *
            </label>
            <input
              data-testid="input-firstname"
              disabled=""
              name="firstname"
              required=""
            />
          </div>
          <div
            class="md:col-span-4"
            data-testid="form-input-lastname"
          >
            <label>
              Firstname
              *
            </label>
            <input
              data-testid="input-lastname"
              disabled=""
              name="lastname"
              required=""
            />
          </div>
          <div
            class="md:col-span-4"
            data-testid="form-input-email"
          >
            <label>
              Email
              *
            </label>
            <input
              data-testid="input-email"
              disabled=""
              name="email"
              required=""
            />
          </div>
          <div
            class="md:col-span-4"
            data-testid="form-input-phone"
          >
            <label>
              Phone
            </label>
            <input
              data-testid="input-phone"
              name="phone"
            />
          </div>
          <div
            class="md:col-span-4"
            data-testid="form-select-address.country"
          >
            <label>
              Country
              *
            </label>
            <select
              data-testid="select-address.country"
              name="address.country"
              required=""
            >
              <option
                value="US"
              >
                United States
              </option>
              <option
                value="CA"
              >
                Canada
              </option>
            </select>
          </div>
          <div
            class="md:col-span-5"
            data-testid="form-input-address.street"
          >
            <label>
              Street
              *
            </label>
            <input
              data-testid="input-address.street"
              name="address.street"
              required=""
            />
          </div>
          <div
            class="md:col-span-3"
            data-testid="form-input-address.apartment"
          >
            <label>
              Apartment
              *
            </label>
            <input
              data-testid="input-address.apartment"
              name="address.apartment"
              required=""
            />
          </div>
          <div
            class="md:col-span-3"
            data-testid="form-input-address.city"
          >
            <label>
              City
              *
            </label>
            <input
              data-testid="input-address.city"
              name="address.city"
              required=""
            />
          </div>
          <div
            class="md:col-span-3"
            data-testid="form-input-address.region"
          >
            <label>
              Region
            </label>
            <input
              data-testid="input-address.region"
              name="address.region"
            />
          </div>
          <div
            class="md:col-span-2"
            data-testid="form-input-address.zip"
          >
            <label>
              Zip
              *
            </label>
            <input
              data-testid="input-address.zip"
              name="address.zip"
              required=""
            />
          </div>
          <button
            class="btn-update md:col-span-full float-right m-2 mt-6 lg:max-w-60 flex flex-1 items-center justify-center rounded-md border border-transparent px-8 py-3 text-base font-medium bg-gray-200 text-gray-800 pointer-events-none"
            disabled=""
          >
            Update
          </button>
        </div>
      </div>
    </form>
  </div>
</body>
`;
